### PR TITLE
Fix-#3296: Patch __array__ method for RPyC proxy objects

### LINKED ISSUE #3296: TypeError: __array__() takes 1 positional argument but 2 were given
### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Modin version** (`modin.__version__`): 0.10.0
- **Python version**: 3.8.10 64-bit
- **Code we can use to reproduce**:

To run the test, clone the repository, install the dependencies, activate the installed virtual environment, and run the command:

```shell
  set MODIN_EXPERIMENTAL=1
  set MODIN_ENGINE=Python
  python -m pytest --simulate-cloud=normal modin/pandas/test/dataframe/test_binary.py::test_math_functions[add-rows-scalar]
```

### Describe the problem

<details>
  <summary>There are two very similar problems that can be solved as follows</summary>

  1. Opening file `rpyc\core\netref.py` ([link](https://github.com/tomerfiliba-org/rpyc/blob/master/rpyc/core/netref.py) on GitHub)
  2. Looking in the file `def __array__` ([link](https://github.com/tomerfiliba-org/rpyc/blob/9aa4d64f49cd85941e4c378c55b8abb1496d06e7/rpyc/core/netref.py#L252) on GitHub)
  3. Replacing this:

      ```py
        def __array__(self):
            # Note that protocol=-1 will only work between python
            # interpreters of the same version.
            return pickle.loads(syncreq(self, consts.HANDLE_PICKLE, -1))
      ```

      For this:

      ```py
        def __array__(self, dtype=None):
            # Note that protocol=-1 will only work between python
            # interpreters of the same version.
            res = pickle.loads(syncreq(self, consts.HANDLE_PICKLE, -1))

            if dtype is not None:
                import numpy
                res = numpy.asarray(res, dtype=dtype)

            return res
      ```

</details>

The first of these problems occurs in 2305 tests:

- modin/pandas/test/dataframe/test_binary.py::test_math_functions[add-rows-scalar]
- modin/pandas/test/dataframe/test_binary.py::test_math_functions[add-rows-series_or_list]
- modin/pandas/test/dataframe/test_binary.py::test_math_functions[add-columns-scalar]

And also in 2302 tests.

The second of these problems occurs in 1637 tests:

- modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0-int_data]
- modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level1,multiindex_level0-int_data]
- modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0,last,first,multiindex_level1-int_data]

And also in 1634 tests.

Perhaps fixing this bug will also solve the problems in the following tests, since the problem is most likely the same:

- modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-no_level-True-3-float_nan_data]
- modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-None-True-3-float_nan_data]
- modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-0-True-3-float_nan_data]

And also in 2433 tests.

### Source code / logs

<details>
    <summary>Part's traceback for first problem:</summary>

    ======= Remote Traceback =======
    E     File  "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\internals \construction.py", line 355, in _homogenize
    E       val = sanitize_array(
    E     File  "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\construct ion.py", line 482, in sanitize_array
    E       subarr = maybe_convert_platform(data)
    E     File  "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\dtypes\ca st.py", line 112, in maybe_convert_platform
    E       values = construct_1d_object_array_from_listlike(values)
    E     File  "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\dtypes\ca st.py", line 1639, in construct_1d_object_array_from_listlike
    E       result[:] = values
    E   TypeError: __array__() takes 1 positional argument but 2 were given

</details>

<details>
    <summary>Part's traceback for second problem:</summary>

    ======= Remote Traceback =======
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\series.py", line 364, in __init__
    E       data = sanitize_array(data, index, dtype, copy, raise_cast_failure=True)
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\construction.py", line 480, in sanitize_array
    E       subarr = _try_cast(data, dtype, copy, raise_cast_failure)
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\construction.py", line 598, in _try_cast
    E       subarr = construct_1d_object_array_from_listlike(subarr)
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\dtypes\cast.py", line 1639, in construct_1d_object_array_from_listlike
    E       result[:] = values
    E   TypeError: __array__() takes 1 positional argument but 2 were given

</details>

<details>
    <summary>Part's traceback for third problem:</summary>

    ======= Remote Traceback =======
    E       self._get_axis(axis).identical(ax)
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\indexes\base.py", line 4485, in identical
    E       self.equals(other)
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\indexes\multi.py", line 3464, in equals
    E       other_values = algos.take_nd(
    E     File "C:\Users\alozovsk\Anaconda3\envs\modin\lib\site-packages\pandas\core\algorithms.py", line 1708, in take_nd
    E       indexer = ensure_int64(indexer, copy=False)
    E     File "pandas\_libs\algos_common_helper.pxi", line 93, in pandas._libs.algos.ensure_int64
    E   TypeError: __array__() takes 1 positional argument but 2 were given 

</details>

### PR DESCRIPTION
**Pull request by Lozovskii-Aleksandr**
## What do these changes do?

- [X] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [X] passes `flake8 modin`
- [X] passes `black --check modin`
- [X] signed commit with `git commit -s` 
- [X] Resolves #3296? 
- [X] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date 

## Solution description

The error was in the module `rpyc.core.netref`. 

 For the solution, it was necessary to patch the `_make_method` method. In the `_make_method` method, in the `name == "__array__ "` block, the `_array__` method must accept two parameters: `self` and `type`. Adding logic from `rpyc_patches.py`, I am correcting this error. Now the `__array__` method accepts the `type` parameter and processes it correctly. More information about the solution is written in https://github.com/modin-project/modin/issues/3296.

### CODE REVIEW

File: modin/experimental/cloud/rpyc_patches.py
Comment (vnlitvinov) [inline modin/experimental/cloud/rpyc_patches.py:16]: I propose to move that `import` to the top of the file

**Review by vnlitvinov — COMMENTED** [review]
Please fix the CI and clean up commit history a little bit.

File: .github/workflows/ci.yml
Comment (vnlitvinov) [inline .github/workflows/ci.yml:551]:
```suggestion
      # When running without parameters, some of the tests fail
```

File: modin/experimental/cloud/rpyc_patches.py
Comment (vnlitvinov) [inline modin/experimental/cloud/rpyc_patches.py:13]:
```suggestion
                    netref.syncreq(self, netref.consts.HANDLE_PICKLE, pickle.HIGHEST_PROTOCOL)
```
let's be explicit in protocol versioning to remove this uncertainty about Python versions' compatibility

**Review by vnlitvinov — COMMENTED** [review]
Changes look good, let's wait until CI passes.

**Review by vnlitvinov — APPROVED** [review]
LGTM when tests pass.

**Comment by YarShev** [comment]
@Lozovskii-Aleksandr , why wasn't any test added to CI to verify the issue resolved?

**Comment by Lozovskii-Aleksandr** [comment]
> @Lozovskii-Aleksandr , why wasn't any test added to CI to verify the issue resolved?

@YarShev, I added one test to `ci.yml` that didn't work before, and after fixing it, it was fixed. It would be possible to add a group of tests that do not have parameters, but it will take a very long time to run. If I'm not mistaken, there are ~40 tests running for 20+ minutes. It would be very difficult to find something faster, since many tests have already failed due to other errors. Therefore, I added only one test.

**Comment by YarShev** [comment]
@Lozovskii-Aleksandr , Shouldn't we also add the tests you mentioned in the issue to CI. I mean these ones:

* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0-int_data]
* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level1,multiindex_level0-int_data]
* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0,last,first,multiindex_level1-int_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-no_level-True-3-float_nan_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-None-True-3-float_nan_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-0-True-3-float_nan_data]

It is said in the issue, your changes might fix the two problems that are similar but not identical.

**Comment by Lozovskii-Aleksandr** [comment]
@YarShev, The tests listed by you are not passing now. There are still errors in them that will need to be corrected. But the error that was corrected in this PR is no longer found in them. 

In the problem, I said that the first and second problems are as similar as possible, and the third is probably related to them. This turned out to be true, since after the correction, the specified error ceased to occur in them.

**Comment by YarShev** [comment]
@Lozovskii-Aleksandr , ok, got it!

**Pull request merged**
